### PR TITLE
[ignore] check symints in dispatcher

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117895
* __->__ #117412

```python
import torch
from timeit import default_timer
from torch.utils.benchmark import Timer

global_setup = """
#include <c10/core/SymInt.h>
"""
setup = """
std::vector<c10::SymInt> size = {c10::SymInt(4), c10::SymInt(10), c10::SymInt(1), c10::SymInt(1), c10::SymInt(1)};
"""

def run(code):
  t = Timer(stmt=code, setup=setup, global_setup=global_setup, language="c++", timer=default_timer)
  print(t.blocked_autorange())

run("""
at::Tensor a = at::empty_symint(size);
""")
```

Before
```
<torch.utils.benchmark.utils.common.Measurement object at 0x1374916c0>
at::Tensor a = at::empty_symint(size);
setup: std::vector<c10::SymInt> size = {c10::SymInt(4), c10::SymInt(10), c10::SymInt(1), c10::SymInt(1), c10::SymInt(1)};
  Median: 271.04 ns
  IQR:    9.96 ns (268.08 to 278.04)
  703 measurements, 1000 runs per measurement, 1 thread
  ```
  
After
```
<torch.utils.benchmark.utils.common.Measurement object at 0x13e6916c0>
at::Tensor a = at::empty_symint(size);
setup: std::vector<c10::SymInt> size = {c10::SymInt(4), c10::SymInt(10), c10::SymInt(1), c10::SymInt(1), c10::SymInt(1)};
  Median: 283.17 ns
  IQR:    6.01 ns (282.71 to 288.72)
  684 measurements, 1000 runs per measurement, 1 thread
```

- We only care about tensors with dim <= 5, bc otherwise we'll need to allocate heap memory to store SizeAndStride anyway! Updating dim 5->6 leads to a consistent ~20ns slowdown with and without the new logic.
-  Every iteration of the loop, we only need to check heap_allocated(), i.e. data_ > MAX_UNREPRESENTABLE_INT. (in the case where we have a SymNode which is non-Singleton that will need another check, but that's fine as most likely its for compile)
- We need to loop through a dynamic number of sizes, so this loop cannot be unrolled.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

### CODE REVIEW

File: aten/src/ATen/core/dispatch/DispatchKeyExtractor.h [hunk 1]
```diff
@@ -94,6 +94,14 @@ namespace detail {
...
       }
     }
+    void operator()(const c10::SymIntArrayRef& xs) {
+      for (const auto& x : xs) {
+        if (x.is_heap_allocated() && x.toSymNode()->singleton_int().has_value()) {
+          ts = ts | DispatchKeySet({DispatchKey::Python, DispatchKey::PythonTLSSnapshot});
+          break;
+        }
+      }
+    }
```
Comment (soulitzer) [inline aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:104]: In the actual impl the current plan is actually to do something like:
```
if (x.is_heap_allocated()) {
    ts = ts | x.toSymNode()->keyset()
}
```

File: torch/csrc/utils/python_arg_parser.cpp [hunk 2]
```diff
@@ -12,6 +12,7 @@
 #include <ATen/ATen.h>
 #include <ATen/PythonTorchFunctionTLS.h>
 #include <ATen/TracerMode.h>
+#include <c10/core/SafePyObject.h>
```
Comment (soulitzer) [inline torch/csrc/utils/python_arg_parser.cpp:15]: To revert
